### PR TITLE
Improve dump lex

### LINKED ISSUE #1650: Improve lexer dump
For now, `-frust-dump-lex` option does not work. It creates `gccrs.lex.dump` with no content because lex dump runs after a crate is parsed.
So I would like to improve lex dump so that lex dump runs at the same time as the parser runs.
I already almost completed modifying the code.

### PR DESCRIPTION
**Pull request by tamaroning**
Fixes #1650

### CODE REVIEW

**Comment by tamaroning** [comment]
Currently, output is like this:
```rust
fn main() {
    let a: i32 = 1 + 2;
    return;
}
```

```
<id=FN_TOK a.rs:1:1 <id=IDENTIFIER, text=main, typehint=unknown a.rs:1:4 <id=LEFT_PAREN a.rs:1:8 <id=RIGHT_PAREN a.rs:1:9 <id=LEFT_CURLY a.rs:1:11 <id=LET a.rs:2:5 <id=IDENTIFIER, text=a, typehint=unknown a.rs:2:9 <id=COLON a.rs:2:10 <id=IDENTIFIER, text=i32, typehint=unknown a.rs:2:12 <id=EQUAL a.rs:2:16 <id=INT_LITERAL, text=1, typehint=pure_decimal a.rs:2:18 <id=PLUS a.rs:2:20 <id=INT_LITERAL, text=2, typehint=pure_decimal a.rs:2:22 <id=SEMICOLON a.rs:2:23 <id=RETURN_TOK a.rs:3:5 <id=SEMICOLON a.rs:3:11 <id=RIGHT_CURLY a.rs:4:1
```

File: gcc/rust/lex/rust-lex.h [hunk 1]
```diff
@@ -139,7 +139,8 @@ class Lexer
 
 public:
   // Construct lexer with input file and filename provided
-  Lexer (const char *filename, RAIIFile input, Linemap *linemap);
+  Lexer (const char *filename, RAIIFile input, Linemap *linemap,
+	 bool dump_lex_enabled);
```
Comment (CohenArthur) [inline gcc/rust/lex/rust-lex.h:143]:
```suggestion
	 bool dump_lex_enabled = false);
```

This allows lexer users who don't care about the dumping to not pass an extra parameter (macros, modules, etc)

File: gcc/rust/lex/rust-lex.h [hunk 2]
```diff
@@ -197,6 +198,9 @@ class Lexer
    * allocating new linemap */
   static const int max_column_hint = 80;
 
+  bool dump_lex_enabled;
+  std::ofstream dump_lex_out;
```
Comment (CohenArthur) [inline gcc/rust/lex/rust-lex.h:202]: This is good. I think the "best" way to go about this would be to keep a reference here, so 
```suggestion
  std::ofstream &dump_lex_out;
```

and to create the `ofstream` in the session manager.

This would allow all "dump operations" to stem from the same file and all "dump filenames" to reside in the same place

File: gcc/rust/lex/rust-lex.cc [hunk 3]
```diff
@@ -24,6 +24,9 @@
 #include "safe-ctype.h"
 
 namespace Rust {
+
+const char *kLexDumpFile = "gccrs.lex.dump";
```
Comment (CohenArthur) [inline gcc/rust/lex/rust-lex.cc:28]: And would allow to move this to the session manager

File: gcc/rust/lex/rust-lex.cc [hunk 4]
```diff
@@ -186,6 +202,50 @@ Lexer::skip_input ()
...
+	      dump_lex_out << tok->token_id_to_str ();
+	      dump_lex_out << (has_text
+				 ? (std::string (", text=") + tok->get_str ()
+				    + std::string (", typehint=")
+				    + std::string (tok->get_type_hint_str ()))
+				 : "")
+			   << " ";
+	      dump_lex_out << get_line_map ()->to_string (loc) << " ";
+	    }
+
```
Comment (CohenArthur) [inline gcc/rust/lex/rust-lex.cc:239]: This should be in a separate function, so that we could have something like this:

```cpp
if (dump_lex_enabled)
    dump_until_n();
   
token_queue.skip(n);
```

File: gcc/rust/lex/rust-lex.h (hunk 2 above)
Comment (CohenArthur) [reply gcc/rust/lex/rust-lex.h:202]: What we could also do here is keep an `Optional<&std::ofstream>` and remove the `dump_lex_enabled` boolean. This way, we simply maybe keep a stream and dump to it when skipping tokens. This would also move the error handling (if the stream fails to open etc) to the session manager, where similar things already happen

File: gcc/rust/lex/rust-lex.cc [hunk 5]
```diff
@@ -186,6 +202,50 @@ Lexer::skip_input ()
   skip_input (0);
 }
 
+void
+Lexer::skip_token (int n)
+{
+  // dump tokens if dump-lex option is enabled
+  if (dump_lex_enabled && !dump_lex_out.fail ())
```
Comment (CohenArthur) [inline gcc/rust/lex/rust-lex.cc:209]: `dump_lex_out.fail()` should not be checked here imo. If it fails to open, it's a compiler error and we should trigger an assertion in the constructor

File: gcc/rust/lex/rust-lex.cc [hunk 6]
```diff
@@ -118,19 +121,30 @@ is_non_decimal_int_literal_separator (char character)
...
   // inform line_table that file is being entered and is in line 1
   if (linemap)
     line_map->start_file (filename, current_line);
+  if (dump_lex_enabled)
+    {
+      dump_lex_out.open (kLexDumpFile);
+      if (dump_lex_out.fail ())
+	{
+	  rust_error_at (Linemap::unknown_location (),
+			 "cannot open %s:%m; ignored", kLexDumpFile);
```
Comment (CohenArthur) [inline gcc/rust/lex/rust-lex.cc:145]: Do we really want to ignore it?

**Review by CohenArthur — COMMENTED** [review]
I'm not sure how I feel about this. I like the idea of having a lexer dump, and you've done some good work :) I like the output format too and I think it's informative.

However, I don't like that the lexer has to take care of opening the dump file, and that the dump file name is defined in `rust-lex.cc` instead of alongside the others in the session-manager. It also feels very "intrusive" at the moment - the dump is being done as parsing is done, and this produces a lot of hard to follow code logic. What other dumps do is that they simply analyze/dump whatever it is they look at *after* that pass has been done in the compiler. Now, I understand that the lexer is different in that aspect: since the queue is buffered, and permanently replaced, we can't simply have the following:

```c++
Lexer lex(...);
Parser<Lexer> parser(lex);

auto crate = parser.parse_crate();

dump_lex(lex);
```

since at this point the lexer will have gone through its entire token queue. So the lexer dump *will* be more intrusive than other dumps, but at the moment it feels *too* intrusive

**Comment by tamaroning** [comment]
@CohenArthur Thank you for your speedy review!
I am a super C++ newbie, so your suggestions help me a lot :)
I will fix them in the next 7 days.

**Comment by CohenArthur** [comment]
@tamaroning thank you :) Feel free to ask any questions, either here or on our Zulip. We'll be happy to help!

File: gcc/rust/lex/rust-lex.cc (hunk 6 above)
Comment (tamaroning) [reply gcc/rust/lex/rust-lex.cc]: I have no idea about that :(

File: gcc/rust/lex/rust-lex.cc (hunk 5 above)
Comment (tamaroning) [reply gcc/rust/lex/rust-lex.cc]: fixed

File: gcc/rust/lex/rust-lex.cc (hunk 4 above)
Comment (tamaroning) [reply gcc/rust/lex/rust-lex.cc]: fixed

File: gcc/rust/lex/rust-lex.cc (hunk 3 above)
Comment (tamaroning) [reply gcc/rust/lex/rust-lex.cc]: fixed

File: gcc/rust/lex/rust-lex.h (hunk 2 above)
Comment (tamaroning) [reply gcc/rust/lex/rust-lex.h]: fixed

File: gcc/rust/lex/rust-lex.h (hunk 1 above)
Comment (tamaroning) [reply gcc/rust/lex/rust-lex.h]: fixed

**Comment by tamaroning** [comment]
Fixed :)

File: gcc/rust/lex/rust-lex.cc [hunk 7]
```diff
@@ -186,6 +188,56 @@ Lexer::skip_input ()
...
+      dump_until_n (n);
+    }
+  else
+    {
+      token_queue.skip (n);
+    }
+}
+
+void
+Lexer::dump_until_n (int n)
```
Comment (CohenArthur) [inline gcc/rust/lex/rust-lex.cc:206]: Ah, so the name I suggested does not make sense anymore since you "either dump or simply skip" in `skip_token (int n)`. So I think this function should be renamed. Something like `void Lexer::dump_and_skip (int n)`

File: gcc/rust/rust-session-manager.cc [hunk 8]
```diff
@@ -439,20 +439,37 @@ Session::compile_crate (const char *filename)
...
 
   // generate crate from parser
   std::unique_ptr<AST::Crate> ast_crate = parser.parse_crate ();
 
+  if (dump_lex_opt.is_some ())
+    {
+      std::ofstream &out = dump_lex_opt.get ();
+      out.close ();
+    }
+
```
Comment (CohenArthur) [inline gcc/rust/rust-session-manager.cc:468]: Is this necessary? The `Optional` should simply go out of scope and destroy its member, in that case the `ofstream` whose destructor takes care of closing the stream?

File: gcc/rust/util/rust-optional.h [hunk 9]
```diff
@@ -194,6 +194,7 @@ template <typename T> class Optional<T &>
 public:
   Optional (const Optional &other) = default;
   Optional (Optional &&other) = default;
+  Optional &operator= (Optional &&other) = default;
```
Comment (CohenArthur) [inline gcc/rust/util/rust-optional.h:197]: Ah, nice catch

**Review by CohenArthur — COMMENTED** [review]
Thanks again for the work :) I think this is in a nice state and we can merge it soon

File: gcc/rust/rust-session-manager.cc (hunk 8 above)
Comment (tamaroning) [reply gcc/rust/rust-session-manager.cc:468]: Oh you are correct. Destructor of ofstream is automatically called and closes the stream.

**Comment by tamaroning** [comment]
Thank you for your quick review @CohenArthur . Fixed.

**Review by CohenArthur — APPROVED** [review]
Great stuff, thanks a lot @tamaroning !

Could you please rebase all commits into a single commit? Then I'll merge your PR. I really appreciate the work you put into this :)

**Comment by tamaroning** [comment]
@CohenArthur Rebased :)

**Review by CohenArthur — APPROVED** [review]
Thank you! And sorry for the delay :)

bors r+
